Isolate logging in tests

abravalheri opened the pull request:
This PR is an attempt to solve issue #170.

I have removed order-dependent test helpers and used unique names instead. Additionally, this PR introduces a new context fixture in an attempt to better isolate the logging usage in the tests.

An error (previously undetected) in the chdir logging is also corrected.

However, the results of the tests are still non-deterministic: in some occasions `caplog.text` is unexpectedly blank as indicated by the CI logs. Any tip/hunch/help on this issue is more then welcome, I couldn't completely figure it out yet.

(Who would say, logging would be the most difficult thing to test??)

Regarding performance: in the CI, the speed improvement is not very noticeable (in part because of the setup time and the system tests in bash that do not run in parallel - running everything inside pytest as suggested in #120 could help).

FlorianWilhelm commented on src/pyscaffold/utils.py:
```diff
@@ -20,6 +20,17 @@
 
 
+@contextmanager
+def _chdir_logginng_context(path, should_log):
```
logginng? Is this a typo or intended?

abravalheri replied:
Typo :)

Thanks for the review!

FlorianWilhelm commented on tests/helpers.py:
```diff
@@ -0,0 +1,25 @@
+    return ''.join(choice(ascii_letters) for _ in xrange(n))
+
+
+def uniqstr():
```
Aren't there any functions in Python's standard module `uuid` that could be used? Just wondering.

abravalheri replied:
UUID could be used, but some extra logic might be required, because I believe python logging required valid identifiers (something like stripping dashes and prepending letters).

If you prefer, I can go in that direction.

FlorianWilhelm replied:
No, I was just wondering if it was considered. I always tend to prefer simpler solution. So if using `uuid` would result in something even simpler but I guess we should keep it as it is right now.

abravalheri replied:
I did some experiments, and it seems that, despite [suggesting it](https://docs.python.org/3.6/library/logging.html#logger-objects), the `module identifier` format is just recommended, not required.

I will try to remove `uniqstr` in favor of `uuid4` and see what happens :)

Thanks for the suggestion.

FlorianWilhelm commented on tests/conftest.py:
```diff
@@ -59,6 +62,38 @@ def logger():
 
 
+@pytest.fixture
+def isolated_logger(logger):
```
For future contributors (and future me ;-) I think it makes sense to describe better what is done here conceptually. As far as I see it you solve the whole singleton logging issue by creating multiple ones with unique ids, right? Somewhere it should be described.

abravalheri commented:
Hi Florian, thanks for the review! Please find my answers in line:

```
This seems to be the final step for complete isolation, right?
Can you yet tell what speedup in the unit-tests we get?
```
Final step \o/.  In the CI the speed up seems to be negligible (something around 1 minute, but I haven't done proper statistics). When running locally, the speed up is more noticeable when we use more than 10 processes, even in machines with a lower number of cores (I am testing in a VM with 2).

I believe this is because we do a lot of IO. However changing dynamically the number of process in xdist is not easy (I would like to have something like `max(10, numcores)` or `numcores + 5` to fully take advantage of the CI) - right now the configuration is to use the number of cores available.

Pytest allow us [to do so](https://docs.pytest.org/en/latest/example/simple.html#dynamically-adding-command-line-options), but not from `conftest`. (I am still investigating if we need an setuptools entrypoint for that or not - if we need I don't think it is worthy the trouble to create a pytest plugin).

```
And has it any more downsides like unreadable logs?
I could imagine that if several unit tests run in parallel you get an confusing output.
```
Writing tests become more difficult, but not that much. (Just when you are using shred resources, like files - I think I fixed the logs now!).
It turns out pytest do a good job organizing the test logs... You can check out the CI ones to get an idea.

```
Can it easily be deactivated if there is any need?
Would be nice to also that that somewhere, how to do it.
```
It would require a custom command line parameter, something similar to the pytest configuration/plugin I have previously mentioned in my first comment. It can be quite difficult to do.

Last comments:
- I might have fixed the non-deterministic behaviour :)
- If we decide to not do parallel testing, I still think some changes in this PR are welcome, and having isolation is something good.

FlorianWilhelm commented:
Thanks @abravalheri, isolation is definitely something we want and the same goes for deterministic behaviour. 

I took a look at the Travis logging and something I noticed is a lot of messages like:
```
tests/api/test_helpers.py::test_merge_rules_just_in_merged Coverage.py warning: Module pyscaffold was previously imported, but not measured (module-not-measured)
Coverage.py warning: Module pyscaffold was previously imported, but not measured (module-not-measured)
Coverage.py warning: Module pyscaffold was previously imported, but not measured (module-not-measured)
```
as well as:
```
Module already imported so cannot be rewritten: coverage
```
Is this related to xdist?

abravalheri commented:
Hi @FlorianWilhelm,

The warning `Module already imported so can not be re-written` at the end of the tests, is something that appears to be quite [common](https://github.com/pytest-dev/pytest-cov/issues/148) when using xdist, and is related to the random order pytest imports its plugins (the solution is something still being [debated](https://github.com/pytest-dev/pytest/issues/935)).

My pragmatic fix for the warning was removing the `-p coverage` from the list of arguments in `setup.cfg`.

The warning `Module pyscaffold was previously imported, but not measured`, seems to be related to importing `pyscaffold` in `conftest.py`. I reckon pytest loads conftest first because of the plugins and configurations that can be there, but then xdist have to load the entire test environment again for each process, setting up the "coverage probes". At this point, `pyscaffold` is already cached in memory.

My pragmatic fix for that is lazy loading everything inside conftest.

abravalheri commented:
@FlorianWilhelm, since it is not easily to "opt-out" of xdist if we include its configuration on `setup.cfg`, I have changed to a `opt-in` approach and added some instructions to the `CONTRIBUTING.md`.

I hope this is a good compromise.

Soon I will provide you some numbers about the real difference in time of using xdist or not.

FlorianWilhelm commented:
@abravalheri, looks cool! Thanks for your work!
